lib/azure/armrest/resource_provider_service.rb

```diff
@@ -1,89 +1,54 @@
 module Azure
   module Armrest
     class ResourceProviderService < ArmrestService
+      extend Memoist
+
       # The provider used in http requests. The default is 'Microsoft.Resources'
       attr_reader :provider
 
-      # The amount of time in seconds to cache certain methods. The default is 24 hours.
-      @cache_time = 24 * 60 * 60
-
-      class << self
-        # Get or set the cache time for all methods. The default is 24 hours.
-        attr_accessor :cache_time
-      end
-
       # Creates and returns a new ResourceProviderService object.
       #
-      # Note that many ResourceProviderService instance methods are cached. You
-      # can set the cache_time for certain methods in the constructor, but keep in
-      # mind that it is a global setting for the class. You can also set this
-      # at the class level if desired. The default cache time is 24 hours.
+      # Note that many ResourceProviderService instance methods are cached.
       #
       # You can also set the provider. The default is 'Microsoft.Resources'.
       #
       def initialize(configuration, options = {})
         super(configuration, 'resourceGroups', 'Microsoft.Resources', options)
-
-        if options[:cache_time]
-          @cache_time = options[:cache_time]
-          self.class.send(:cache_time=, @cache_time)
-        end
       end
 
       # List all the providers for the current subscription. The results of
       # this method are cached.
       #
       def list
-        _list.map{ |hash| Azure::Armrest::ResourceProvider.new(hash) }
-      end
-
-      # This is split out for the cache_method feature.
-      #
-      def _list
         response = rest_get(build_url)
-        JSON.parse(response)["value"]
+        resources = JSON.parse(response)["value"]
+        resources.map{ |hash| Azure::Armrest::ResourceProvider.new(hash) }
       end
-
-      private :_list
-
-      # Cannot directly cache list method, because Marshal used by cache_method
-      # cannot dump anonymous classes, including the ones derived from
-      # Azure::Armrest::BaseModel.
-      #
-      # The same applies to other cached methods in this class.
-      cache_method(:_list, cache_time)
+      memoize :list
```

Comment:
> cache_method cannot cache Armrest model instances, and it is the reason we have to cache the underline hash object.

That's because cache-method use `marshal_dump` which can't dump singletons.  Memoist just uses a normal hash cache.